scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala

```diff
@@ -139,12 +139,13 @@ class BestFirstSearch(
       depth: Int = 0,
       maxCost: Int = Integer.MAX_VALUE
   ): State = {
-    val Q = new mutable.PriorityQueue[State]()
+    val Q = new mutable.PriorityQueue[(Int, State)]()
```

Comment:
I just noticed the commit here https://github.com/scalameta/scalafmt/commit/2cd7a731259df9bfbb76d3b3cc4ed1f5d24f076e#diff-0944a127ca40d4a12020bbe4e69623cb

The ordering comparison method is quite performance sensitive so I think it's better to try and avoid redundant allocations and implement the ordering logic on `Integer.compare(...)` like it was done before.

We might also want to encode the generation number as a field on `State` and update the ordering logic.

Reply:
> I just noticed the commit here [2cd7a73#diff-0944a127ca40d4a12020bbe4e69623cb](https://github.com/scalameta/scalafmt/commit/2cd7a731259df9bfbb76d3b3cc4ed1f5d24f076e#diff-0944a127ca40d4a12020bbe4e69623cb)
> 
> The ordering comparison method is quite performance sensitive so I think it's better to try and avoid redundant allocations and implement the ordering logic on `Integer.compare(...)` like it was done before.

i will try to figure out how to profile the code, but I'm fairly certain that using `Integer.valueOf` (which turned primitives into boxed objects) must have had a cost.
> 
> We might also want to encode the generation number as a field on `State` and update the ordering logic.

let's try.

p.s. one unrelated question, regarding performance. there's a memo lookup there, I'm guessing the dp part of the algorithm, on the column/indentation of the intermediate solution. has anyone tried to observe that if a given solution works for some starting column but has some slack on the right before it hits maxColumn, it can be shifted without recomputing? is that even true?

Reply:
`Integer.compare` doesn't box primitives
```java
    public static int compare(int x, int y) {
```

> has anyone tried to observe that if a given solution works for some starting column but has some slack on the right before it hits maxColumn, it can be shifted without recomputing? is that even true?

Nobody has looked into that from what I know

Reply:
> `Integer.compare` doesn't box primitives
> 
> ```java
>     public static int compare(int x, int y) {
> ```

of course! but this does (https://github.com/scalameta/scalafmt/commit/2cd7a731259df9bfbb76d3b3cc4ed1f5d24f076e#diff-0944a127ca40d4a12020bbe4e69623cb):
```
  val costCompare = Integer.valueOf(-this.cost).compareTo(-that.cost)

  // from Integer.java
    public static Integer valueOf(int i) {
        if (i >= IntegerCache.low && i <= IntegerCache.high)
            return IntegerCache.cache[i + (-IntegerCache.low)];
        return new Integer(i);
    }
```

i will look into performance of `State` comparisons later, but for now rather than putting `gen` into `State`, what about just implementing it as a list of priority queues instead? allocating a new queue and comparing values in it is likely cheaper than putting a value in `State` and comparing all `State`s.

Reply:
Aha, yes indeed :)

There's no good reason to do `Integer.valueOf(primitive).compareTo(..)`

Reply:
> what about just implementing it as a list of priority queues instead? 

Sounds good to me 👍